examples/chatbot/src/websocket.rs

```diff
@@ -0,0 +1,175 @@
+            tokio_tungstenite::MaybeTlsStream<tokio::net::TcpStream>,
+        >,
+        eyre::Error,
+    > {
+        tracing::info!("connecting to twitch");
+        let config = tungstenite::protocol::WebSocketConfig::default();
+        let (socket, _) =
+            tokio_tungstenite::connect_async_with_config(&self.connect_url, Some(config), false)
+                .await
+                .wrap_err("Can't connect")?;
+
+        Ok(socket)
+    }
+
+    /// Run the websocket subscriber
+    #[tracing::instrument(name = "subscriber", skip_all, fields())]
+    pub async fn run<Fut>(
+        mut self,
+        mut event_fn: impl FnMut(Event, types::Timestamp) -> Fut,
+    ) -> Result<(), eyre::Error>
+    where
+        Fut: std::future::Future<Output = ()>,
+    {
+        // Establish the stream
+        let mut s = self
+            .connect()
+            .await
+            .context("when establishing connection")?;
+        // Loop over the stream, processing messages as they come in.
+        loop {
+            tokio::select!(
+            Some(msg) = futures::StreamExt::next(&mut s) => {
+                let span = tracing::debug_span!("message received", raw_message = ?msg);
+                let msg = match msg {
+                    Err(tungstenite::Error::Protocol(
+                        tungstenite::error::ProtocolError::ResetWithoutClosingHandshake,
+                    )) => {
+                        tracing::warn!(
+                            "connection was sent an unexpected frame or was reset, reestablishing it"
+                        );
+                        s = self
+                            .connect().instrument(span)
+                            .await
+                            .context("when reestablishing connection")?;
+                        continue
+                    }
+                    _ => msg.context("when getting message")?,
+                };
+                self.process_message(msg, &mut event_fn).instrument(span).await?
+            })
+        }
+    }
+
+    /// Process a message from the websocket
+    pub async fn process_message<Fut>(
+        &mut self,
+        msg: tungstenite::Message,
+        event_fn: &mut impl FnMut(Event, types::Timestamp) -> Fut,
+    ) -> Result<(), eyre::Report>
+    where
+        Fut: std::future::Future<Output = ()>,
+    {
+        match msg {
+            tungstenite::Message::Text(s) => {
+                tracing::trace!("{s}");
+                // Parse the message into a [twitch_api::eventsub::EventsubWebsocketData]
+                match Event::parse_websocket(&s)? {
+                    EventsubWebsocketData::Welcome {
+                        payload: WelcomePayload { session },
+                        ..
+                    }
+                    | EventsubWebsocketData::Reconnect {
+                        payload: ReconnectPayload { session },
+                        ..
+                    } => {
```

Comment:
We never reconnect to the provided URL. Should we handle this at all?

Reply:
lets do it, it's an easy fix

Reply:
oh actually, we do, if there's a reconnect url we connect to it